File: src/main/java/cc/mirukuneko/realtrainmodrenewed/RealTrainModRenewedBlocks.java
package cc.mirukuneko.realtrainmodrenewed;

import cc.mirukuneko.realtrainmodrenewed.block.BallastBlock;
import cc.mirukuneko.realtrainmodrenewed.block.CrossingGateBlock;
import cc.mirukuneko.realtrainmodrenewed.block.InstalledObjectBlock;
import cc.mirukuneko.realtrainmodrenewed.block.LargeRailCoreBlock;
import cc.mirukuneko.realtrainmodrenewed.block.MarkerBlock;
import cc.mirukuneko.realtrainmodrenewed.block.RailCollisionBlock;
import cc.mirukuneko.realtrainmodrenewed.block.ScriptBlock;
import cc.mirukuneko.realtrainmodrenewed.block.SignalRemoteBlock;
import cc.mirukuneko.realtrainmodrenewed.block.SignalStateBlock;
import cc.mirukuneko.realtrainmodrenewed.block.TrainDetectorBlock;
import net.minecraft.world.level.block.SoundType;
import net.minecraft.world.level.block.state.BlockBehaviour;
import net.neoforged.neoforge.registries.DeferredBlock;
import net.neoforged.neoforge.registries.DeferredRegister;

public class RealTrainModRenewedBlocks {
    public static final DeferredRegister.Blocks BLOCKS = DeferredRegister.createBlocks(RealTrainModRenewed.MODID);

    public static final DeferredBlock<CrossingGateBlock> CROSSING_GATE
        = BLOCKS.registerBlock("crossing_gate", CrossingGateBlock::new,
            () -> BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.0F, 6.0F).noOcclusion());
    public static final DeferredBlock<MarkerBlock> MARKER
        = BLOCKS.registerBlock("marker", props -> new MarkerBlock(false, props),
            () -> BlockBehaviour.Properties.of().sound(SoundType.STONE).strength(1.0F, 1.0F).noOcclusion().noCollision());
    public static final DeferredBlock<MarkerBlock> MARKER_SWITCH
        = BLOCKS.registerBlock("marker_switch", props -> new MarkerBlock(true, props),
            () -> BlockBehaviour.Properties.of().sound(SoundType.STONE).strength(1.0F, 1.0F).noOcclusion().noCollision());

    /** 道床ブロック（レールと独立した物理ブロック） */
    public static final DeferredBlock<BallastBlock> BALLAST
        = BLOCKS.registerBlock("ballast", BallastBlock::new,
            () -> BlockBehaviour.Properties.of()
                .sound(SoundType.GRAVEL)
                .strength(0.6F, 3.0F)
                .noOcclusion()
                .isSuffocating((s, g, p) -> false)
                .isViewBlocking((s, g, p) -> false));

    /** レールコアブロック（起点1個のみ、MQOモデル描画を担当） */
    public static final DeferredBlock<LargeRailCoreBlock> LARGE_RAIL_CORE
        = BLOCKS.registerBlock("large_rail_core", LargeRailCoreBlock::new,
            () -> BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(0.5F, 6.0F).noOcclusion());

    /** レール当たり判定ブロック（非表示・薄い） */
    public static final DeferredBlock<RailCollisionBlock> RAIL_COLLISION
        = BLOCKS.registerBlock("rail_collision", RailCollisionBlock::new,
            () -> BlockBehaviour.Properties.of()
                .sound(SoundType.METAL)
                .strength(0.1F, 0.1F)
                .noOcclusion()
                .isSuffocating((s, g, p) -> false)
                .isViewBlocking((s, g, p) -> false));

    public static final DeferredBlock<InstalledObjectBlock> INSTALLED_OBJECT
        = BLOCKS.registerBlock("installed_object", InstalledObjectBlock::new,
            () -> BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(0.4F, 2.0F).noOcclusion());

    public static final DeferredBlock<SignalRemoteBlock> SIGNAL_RECEIVER
        = BLOCKS.registerBlock("signal_receiver", props -> new SignalRemoteBlock(props, SignalRemoteBlock.Mode.RECEIVER),
            () -> BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.0F, 6.0F));

    public static final DeferredBlock<SignalRemoteBlock> SIGNAL_CHANGER
        = BLOCKS.registerBlock("signal_changer", props -> new SignalRemoteBlock(props, SignalRemoteBlock.Mode.CHANGER),
            () -> BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.0F, 6.0F));

    public static final DeferredBlock<SignalRemoteBlock> SIGNAL_VALUE_RECEIVER
        = BLOCKS.registerBlock("signal_value_receiver", props -> new SignalRemoteBlock(props, SignalRemoteBlock.Mode.VALUE_INPUT),
            () -> BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.0F, 6.0F));

    public static final DeferredBlock<TrainDetectorBlock> TRAIN_DETECTOR
        = BLOCKS.registerBlock("train_detector", TrainDetectorBlock::new,
            () -> BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.2F, 6.0F));

    public static final DeferredBlock<SignalStateBlock> SIGNAL_STATE
        = BLOCKS.registerBlock("signal_state", SignalStateBlock::new,
            () -> BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.2F, 6.0F));

    public static final DeferredBlock<ScriptBlock> SCRIPT_BLOCK
        = BLOCKS.registerBlock("script_block", ScriptBlock::new,
            () -> BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.2F, 6.0F));
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/RealTrainModRenewedBlocks.kt
package cc.mirukuneko.realtrainmodrenewed

import cc.mirukuneko.realtrainmodrenewed.block.*
import net.minecraft.world.level.block.SoundType
import net.minecraft.world.level.block.state.BlockBehaviour
import net.neoforged.neoforge.registries.DeferredBlock
import net.neoforged.neoforge.registries.DeferredRegister
import java.util.function.Function
import java.util.function.Supplier

object RealTrainModRenewedBlocks {
    @JvmField
    val BLOCKS: DeferredRegister.Blocks = DeferredRegister.createBlocks(RealTrainModRenewed.MODID)

    @JvmField val CROSSING_GATE: DeferredBlock<CrossingGateBlock> = BLOCKS.registerBlock("crossing_gate", Function { props: BlockBehaviour.Properties -> CrossingGateBlock(props) }, Supplier { BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.0f, 6.0f).noOcclusion() })
    @JvmField val MARKER: DeferredBlock<MarkerBlock> = BLOCKS.registerBlock("marker", Function { props: BlockBehaviour.Properties -> MarkerBlock(false, props) }, Supplier { BlockBehaviour.Properties.of().sound(SoundType.STONE).strength(1.0f, 1.0f).noOcclusion().noCollision() })
    @JvmField val MARKER_SWITCH: DeferredBlock<MarkerBlock> = BLOCKS.registerBlock("marker_switch", Function { props: BlockBehaviour.Properties -> MarkerBlock(true, props) }, Supplier { BlockBehaviour.Properties.of().sound(SoundType.STONE).strength(1.0f, 1.0f).noOcclusion().noCollision() })
    @JvmField val BALLAST: DeferredBlock<BallastBlock> = BLOCKS.registerBlock("ballast", Function { props: BlockBehaviour.Properties -> BallastBlock(props) }, Supplier { BlockBehaviour.Properties.of().sound(SoundType.GRAVEL).strength(0.6f, 3.0f).noOcclusion().isSuffocating { _, _, _ -> false }.isViewBlocking { _, _, _ -> false } })
    @JvmField val LARGE_RAIL_CORE: DeferredBlock<LargeRailCoreBlock> = BLOCKS.registerBlock("large_rail_core", Function { props: BlockBehaviour.Properties -> LargeRailCoreBlock(props) }, Supplier { BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(0.5f, 6.0f).noOcclusion() })
    @JvmField val RAIL_COLLISION: DeferredBlock<RailCollisionBlock> = BLOCKS.registerBlock("rail_collision", Function { props: BlockBehaviour.Properties -> RailCollisionBlock(props) }, Supplier { BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(0.1f, 0.1f).noOcclusion().isSuffocating { _, _, _ -> false }.isViewBlocking { _, _, _ -> false } })
    @JvmField val INSTALLED_OBJECT: DeferredBlock<InstalledObjectBlock> = BLOCKS.registerBlock("installed_object", Function { props: BlockBehaviour.Properties -> InstalledObjectBlock(props) }, Supplier { BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(0.4f, 2.0f).noOcclusion() })
    @JvmField val SIGNAL_RECEIVER: DeferredBlock<SignalRemoteBlock> = BLOCKS.registerBlock("signal_receiver", Function { props: BlockBehaviour.Properties -> SignalRemoteBlock(props, SignalRemoteBlock.Mode.RECEIVER) }, Supplier { BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.0f, 6.0f) })
    @JvmField val SIGNAL_CHANGER: DeferredBlock<SignalRemoteBlock> = BLOCKS.registerBlock("signal_changer", Function { props: BlockBehaviour.Properties -> SignalRemoteBlock(props, SignalRemoteBlock.Mode.CHANGER) }, Supplier { BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.0f, 6.0f) })
    @JvmField val SIGNAL_VALUE_RECEIVER: DeferredBlock<SignalRemoteBlock> = BLOCKS.registerBlock("signal_value_receiver", Function { props: BlockBehaviour.Properties -> SignalRemoteBlock(props, SignalRemoteBlock.Mode.VALUE_INPUT) }, Supplier { BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.0f, 6.0f) })
    @JvmField val TRAIN_DETECTOR: DeferredBlock<TrainDetectorBlock> = BLOCKS.registerBlock("train_detector", Function { props: BlockBehaviour.Properties -> TrainDetectorBlock(props) }, Supplier { BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.2f, 6.0f) })
    @JvmField val SIGNAL_STATE: DeferredBlock<SignalStateBlock> = BLOCKS.registerBlock("signal_state", Function { props: BlockBehaviour.Properties -> SignalStateBlock(props) }, Supplier { BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.2f, 6.0f) })
    @JvmField val SCRIPT_BLOCK: DeferredBlock<ScriptBlock> = BLOCKS.registerBlock("script_block", Function { props: BlockBehaviour.Properties -> ScriptBlock(props) }, Supplier { BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(1.2f, 6.0f) })
}
